zerver/lib/message.py

```diff
@@ -344,3 +347,86 @@ def render_markdown(message, content, realm=None, realm_alert_words=None, messag
+            for log_entry in stream_subscription_logs:
+                if len(stream_messages) > 0:
+                    if log_entry.event_type == 'subscription_deactivated':
+                        for stream_message in stream_messages:
+                            if stream_message['pub_date'] < log_entry.event_time:
+                                store_user_message_to_insert(stream_message)
+                            else:
+                                break
+                    else:
+                        initial_msg_count = len(stream_messages)
+                        for i, stream_message in enumerate(stream_messages):
+                            if stream_message['pub_date'] >= log_entry.event_time:
+                                stream_messages = stream_messages[i:]
+                                break
+                        final_msg_count = len(stream_messages)
+                        if initial_msg_count == final_msg_count:
+                            if stream_messages[-1]['pub_date'] < log_entry.event_time:
+                                stream_messages = []
+            if len(stream_messages) > 0:
+                for stream_message in stream_messages:
+                    store_user_message_to_insert(stream_message)
+
+    all_subs = Subscription.objects.select_related('recipient').filter(
+        user_profile=user_profile).values('recipient')
+
+    recipient_ids = [sub['recipient'] for sub in all_subs]
+    all_stream_msgs = list(Message.objects.select_related().filter(
+        recipient__id__in=recipient_ids,
+        recipient__type=Recipient.STREAM,
+        id__gt=user_profile.last_active_message_id).order_by('pub_date').values(
+        'id', 'pub_date', 'recipient__type_id'))
+    already_created_ums = list(UserMessage.objects.select_related().filter(
+        user_profile=user_profile,
+        message__recipient__type=Recipient.STREAM,
+        message__id__gt=user_profile.last_active_message_id).values(
+        'message__id'))
+    already_created_ums = [msg['message__id'] for msg in already_created_ums]
+
+    # Filter those messages for which UserMessage rows have been already created
+    all_stream_msgs = [msg for msg in all_stream_msgs
+                       if msg['id'] not in already_created_ums]
+
+    # For Stream messages we need to check messages against data from
+    # RealmAuditLog for visibility to user.
+    stream_messages = defaultdict(list)  # type: DefaultDict[int, List]
+    for msg in all_stream_msgs:
+        stream_messages[msg['recipient__type_id']].append(msg)
+    stream_ids = list(stream_messages.keys())
+    events = ['subscription_created', 'subscription_deactivated', 'subscription_activated']
+    subscription_logs = list(RealmAuditLog.objects.filter(
+        modified_user=user_profile,
+        modified_stream__id__in=stream_ids,
+        event_type__in=events).order_by('event_time'))
```

Comment:
One thing I notice when looking at this is that our life would be a lot simpler if we recorded in RealmAuditLog the maximum ID in the `Message` table at the time of the RealmAuditLog event from effectively `select max(message_id) from zerver_message`.  I think it would be not too hard to modify the `RealmAuditLog` code to record that in a new column that's just used for Subscription objects.  If you like that idea, can you submit a separate PR to add that?  

I backed out the just-merged RealmAuditLog changes related to Subscription to only be present in a branch `timabbott/tabbott-realmauditlog` in my repo so we can do this by just changing the initial migrations for RealmAuditLog rather than adding a more complicated double migration.